pkg/server/delta/v3/server.go

```diff
@@ -44,6 +132,340 @@ func NewServer(ctx context.Context, config cache.ConfigWatcher, callbacks Callba
+			}
+			nonce, err := send(resp, resource.SecretType)
+			if err != nil {
+				return err
+			}
+			values.deltaSecretNonce = nonce
+			setState(resp, resource.SecretType)
+		case resp, more := <-values.deltaRuntimes:
+			if !more {
+				return status.Errorf(codes.Unavailable, "runtimes watch failed")
+			}
+			nonce, err := send(resp, resource.RuntimeType)
+			if err != nil {
+				return err
+			}
+			values.deltaRuntimeNonce = nonce
+			setState(resp, resource.RuntimeType)
+		case resp, more := <-values.deltaResponses:
+			if more {
+				if resp == deltaErrorResponse {
+					return status.Errorf(codes.Unavailable, "delta resource watch failed")
+				}
+				typeURL := resp.GetDeltaRequest().TypeUrl
+				nonce, err := send(resp, typeURL)
+				if err != nil {
+					return err
+				}
+				values.deltaNonces[typeURL] = nonce
+				setState(resp, typeURL)
+			}
+		case req, more := <-reqCh:
+			// input stream ended or errored out
+			if !more {
+				return nil
+			}
+			if req == nil {
+				return status.Errorf(codes.Unavailable, "empty request")
+			}
+
+			// Log out our error detail from  envoyif we get one but don't do anything crazy here yet
+			// TODO: embed a logger in the server so we can be a bit more verbose when needed
+			if req.ErrorDetail != nil {
+				return status.Errorf(codes.Code(req.ErrorDetail.GetCode()), "received error from xDS client: %s", req.ErrorDetail.GetMessage())
+			}
+
+			// node field in discovery request is delta-compressed
+			// nonces can be reused across streams; we verify nonce only if nonce is not initialized
+			var nonce string
+			if req.Node != nil {
+				node = req.Node
+				nonce = req.GetResponseNonce()
+			} else {
+				req.Node = node
+				// If we have no nonce, i.e. this is the first request on a delta stream, set one
+				nonce = strconv.FormatInt(streamNonce, 10)
+			}
+
+			// type URL is required for ADS but is implicit for xDS
+			if defaultTypeURL == resource.AnyType {
+				if req.TypeUrl == "" {
+					return status.Errorf(codes.InvalidArgument, "type URL is required for ADS")
+				}
+			} else if req.TypeUrl == "" {
+				req.TypeUrl = defaultTypeURL
+			}
+
+			state := values.deltaStreamStates[req.GetTypeUrl()]
+			// If this is empty we can assume this is the first time state
+			// is being set on this stream for this resource type
+			if state.ResourceVersions == nil {
+				state.ResourceVersions = make(map[string]string)
+			}
+
+			// We are in the wildcard mode if the first request of a particular type has an empty subscription list
+			var found bool
+			if state.IsWildcard, found = isWildcard[req.TypeUrl]; !found {
+				state.IsWildcard = len(req.GetResourceNamesSubscribe()) == 0
+				isWildcard[req.TypeUrl] = state.IsWildcard
+			}
+
+			if sub := req.GetResourceNamesSubscribe(); len(sub) > 0 {
+				s.subscribe(sub, state.ResourceVersions)
+			}
+			for r, v := range req.InitialResourceVersions {
+				state.ResourceVersions[r] = v
+			}
```

Comment:
Are we assuming that `InitialResourceVersions` is only set on the first request? What happens if it remains set?

Reply:
I think we also need to account for a reconnect, if a stream dies and then reinitiates that could also be full. I'm wondering if we should add a check here to see if it's populated or not?
https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/discovery.proto#envoy-api-msg-deltadiscoveryrequest

Reply:
Yeah we need to read this to be able to resume sessions, I'm just wondering if we should take care to only try to read this on the first request, as otherwise we'd be resetting resource versions to this value? Maybe error out if this is set for subsequent requests for a given type url, since this seems to be in violation with the spec

Reply:
That sounds good to me. I'll go ahead and make that change